c-wrapper/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+/// - **OR** a pointer to a valid buffer of initialized data of at
+/// least `nonce_size` bytes, which will not be mutated for the duration
+/// of this function call.
+///
+/// It is the caller's responsibility to ensure that `out_session` is
+/// not a null pointer.
+#[no_mangle]
+pub unsafe extern "C" fn open_challenge_response_session(
+    base_url: *const libc::c_char,
+    nonce_size: libc::size_t,
+    nonce: *const u8,
+    out_session: *mut *mut ChallengeResponseSession,
+) -> VeraisonResult {
+    // We have to trust the caller's char* ptr.
+    let url_str: &str = {
+        let url_cstr = CStr::from_ptr(base_url);
+        url_cstr.to_str().unwrap()
+    };
+
+    // Make a Nonce variant according to the given nonce_size and nonce arguments. If the nonce is null,
+    // this implies the Nonce::Size() variant, otherwise it's the Nonce::Value() variant.
+    let nonce_converted: Nonce = {
+        if nonce.is_null() {
+            // Null pointer implies a request for the server to generate the nonce
+            // of the given size. The size is also permitted to be zero, in which case
+            // the server will choose the size as well as generating the nonce.
+            Nonce::Size(nonce_size)
+        } else {
+            // Non-null pointer means we are making a Nonce::Value variant of the
+            // given size. We have to trust the caller's pointer here.
+            let bytes = slice::from_raw_parts(nonce, nonce_size);
+            Nonce::Value(Vec::from(bytes))
+        }
+    };
+
+    // Establish the client session.
+    let cr = ChallengeResponseBuilder::new()
+        .with_base_url(String::from(url_str))
+        .build();
+
+    // Early return on error. The idiom here is slightly odd, and it arguably would be better to use
+    // is_err() and expect_err(), but these require the underlying types to implement the Debug trait,
+    // which they don't.
+    match cr {
+        Ok(_) => {}
+        Err(e) => {
+            // Early return with a stub session containing the error message.
+            return stub_session_from_error(&e, out_session);
+        }
+    }
+
+    // This now won't panic because we dealt with errors by early return above.
+    let cr = cr.unwrap();
+
+    let newsession = cr.new_session(&nonce_converted);
+
+    match newsession {
+        Ok(_) => {}
+        Err(e) => {
+            // Early return with a stub session containing the error message.
+            return stub_session_from_error(&e, out_session);
+        }
+    }
+
+    // This now won't panic because we dealt with errors by early return above.
+    let (session_uri, session) = newsession.unwrap();
+
+    let session_nonce = session.nonce().to_vec();
+    let session_accept_types = session.accept().to_vec();
+
+    // Map the Rust Strings to CString objects for the accept types.
+    let media_type_cstrings = session_accept_types
+        .iter()
+        .map(|s| CString::new(s.as_str()).unwrap())
```

Comment:
Could the unwraps on `CString::new` be avoided throughout this function? Or is it not possible to have null bytes in the incoming `String`?

Reply:
I think that the strings we treat with `unwrap()` are either strings that have been parsed from JSON responses from the server, or they are strings coming from C world that are nul-terminated. In either case, it's hard to see how those functions would fail in practice, given how much it would complicate the code to try to handle those failures. So it was a judgement call between the potential for bugs with the code as written, and the potential for bugs in the code that would result from trying to handle each conversion error individually, and figure out how to respond to the caller in each case.

Reply:
Yeah, that's fair! Thanks for confirming/explaining